### PR TITLE
Improve unit tests

### PR DESCRIPTION
**Pull request by jpraet**
* cleanup tests and use AssertJ
* some minor code changes discovered during testing: 
  * InputValidationIssues.requiredInput should not have a value param as it is always null
  * Establishment number -> Establishment unit number
  * preceed -> precede
  * fix InputValidationIssue equals and hashcode
  * add .toString() methods

### CODE REVIEW

File: belgif-rest-problem/src/test/java/io/github/belgif/rest/problem/validation/RefDataCollectionValidatorTest.java [hunk 1]
```diff
@@ -2,66 +2,38 @@
...
+                .validate()).isEmpty();
     }
 
     @Test
-    void validateNullInput() {
-        assertThrows(NullPointerException.class,
-                () -> new RefDataCollectionValidator<>(null, Collections.singletonList("ok")),
-                "Should have thrown exception");
+    void nokNullAllowedRefData() {
+        assertThatNullPointerException().isThrownBy(
```
Comment (clone1612) [inline belgif-rest-problem/src/test/java/io/github/belgif/rest/problem/validation/RefDataCollectionValidatorTest.java:23]: Maybe check exception message here as well?

File: belgif-rest-problem/src/test/java/io/github/belgif/rest/problem/validation/RefDataCollectionValidatorTest.java [hunk 2]
```diff
@@ -2,66 +2,38 @@
...
     }
 
     @Test
-    void validateNullIn() {
-        assertThrows(NullPointerException.class,
-                () -> new RefDataCollectionValidator<>(new Input<>(null, "refData", "ok"),
-                        Collections.singletonList("ok")),
-                "Should have thrown exception");
+    void nokEmptyAllowedRefData() {
+        assertThatIllegalStateException().isThrownBy(
```
Comment (clone1612) [inline belgif-rest-problem/src/test/java/io/github/belgif/rest/problem/validation/RefDataCollectionValidatorTest.java:29]: Maybe check exception message here as well?

File: belgif-rest-problem/src/test/java/io/github/belgif/rest/problem/validation/RefDataPredicateValidatorTest.java [hunk 3]
```diff
@@ -2,56 +2,29 @@
...
-                "Should have thrown exception");
-    }
-
-    @Test
-    void validateNullAllowedRefData() {
-        assertThrows(NullPointerException.class,
-                () -> new RefDataPredicateValidator<>(new Input<>(QUERY, "refData", "ok"), null),
-                "Should have thrown exception");
+    void nokNullAllowedRefData() {
+        assertThatNullPointerException().isThrownBy(
```
Comment (clone1612) [inline belgif-rest-problem/src/test/java/io/github/belgif/rest/problem/validation/RefDataPredicateValidatorTest.java:20]: Maybe check exception message here as well?

File: belgif-rest-problem/src/test/java/io/github/belgif/rest/problem/validation/RequestValidatorTest.java [hunk 4]
```diff
@@ -19,302 +17,432 @@
...
+    }
+
+    @Test
+    void zeroOrAllOfValidAll() {
+        Input<?>[] inputs = { Input.query("a", "ok"), Input.query("b", "ok") };
+        assertValid(new RequestValidator().zeroOrAllOf(inputs));
+    }
+
+    @Test
+    void zeroOrallOfInvalid() {
```
Comment (clone1612) [inline belgif-rest-problem/src/test/java/io/github/belgif/rest/problem/validation/RequestValidatorTest.java:201]: Nitpick: all should be capitalized

File: belgif-rest-problem/src/test/java/io/github/belgif/rest/problem/validation/RequestValidatorTest.java [hunk 5]
```diff
@@ -19,302 +17,432 @@
 import io.github.belgif.rest.problem.api.InputValidationIssues;
 
 class RequestValidatorTest {
```
Comment (clone1612) [inline belgif-rest-problem/src/test/java/io/github/belgif/rest/problem/validation/RequestValidatorTest.java:19]: There doesn't seem to be a test which chains multiple validations and verifies they are all checked (i.e. validation does not fail fast)?

File: belgif-rest-problem/src/test/java/io/github/belgif/rest/problem/validation/TimeReflectionUtilTest.java [hunk 6]
```diff
@@ -49,20 +49,27 @@ void nokWithInvalidType() {
                 "Should have thrown exception");
     }
```
Comment (clone1612) [inline belgif-rest-problem/src/test/java/io/github/belgif/rest/problem/validation/TimeReflectionUtilTest.java:51]: This test class still uses JUnit assertions for validating an exception was thrown

File: belgif-rest-problem/src/test/java/io/github/belgif/rest/problem/validation/ZeroOrAllOfValidatorTest.java [hunk 7]
```diff
@@ -1,59 +1,36 @@
...
 import java.util.List;
 
 import org.junit.jupiter.api.Test;
 
 import io.github.belgif.rest.problem.api.Input;
-import io.github.belgif.rest.problem.api.InputValidationIssue;
 import io.github.belgif.rest.problem.api.InputValidationIssues;
 
 class ZeroOrAllOfValidatorTest {
     private ZeroOrAllOfValidator tested;
```
Comment (clone1612) [inline belgif-rest-problem/src/test/java/io/github/belgif/rest/problem/validation/ZeroOrAllOfValidatorTest.java:14]: This field is no longer used

File: belgif-rest-problem/src/main/java/io/github/belgif/rest/problem/api/InputValidationIssues.java [hunk 8]
```diff
@@ -108,10 +108,10 @@ public static InputValidationIssue rejectedInput(InEnum in, String name, Object
                 .in(in, name, value);
     }
```
Comment (clone1612) [inline belgif-rest-problem/src/main/java/io/github/belgif/rest/problem/api/InputValidationIssues.java:110]: The class also contains `ISSUE_TYPE_UNKNOWN_INPUT` but it isn't used in a method nor does it have a test. Is that normal?

**Review by clone1612 — CHANGES_REQUESTED** [review]
A very nice increase in test quality/coverage, I've added some minor remarks.

File: belgif-rest-problem/src/test/java/io/github/belgif/rest/problem/api/InputValidationIssuesTest.java [hunk 9]
```diff
@@ -0,0 +1,303 @@
...
+
+import java.time.LocalDate;
+import java.time.OffsetDateTime;
+import java.time.OffsetTime;
+import java.util.Arrays;
+import java.util.List;
+
+import org.junit.jupiter.api.Test;
+
+class InputValidationIssuesTest {
```
[the remark this answers is not in the document]
Comment (jpraet) [reply belgif-rest-problem/src/test/java/io/github/belgif/rest/problem/api/InputValidationIssuesTest.java:13]: You mean something like this?

```java
        assertThat(InputValidationIssues.schemaViolation(InEnum.BODY, "test", "value", "detail"))
                .isEqualTo(new InputValidationIssue(ISSUE_TYPE_SCHEMA_VIOLATION,
                        "Input value is invalid with respect to the schema")
                                .detail("detail")
                                .in(InEnum.BODY, "test", "value"));
```

My concern with that is that the test code would then basically do the same as the production code: construct the expected InputValidationIssue. It is likely that such test would be created starting by copying the production code into the test, and potentially just duplicating a bug in the test?

[the remark this answers is not in the document]
Comment (clone1612) [reply belgif-rest-problem/src/test/java/io/github/belgif/rest/problem/api/InputValidationIssuesTest.java:13]: That is indeed a valid concern, what about using the no-args constructor with the setters (not the builder methods)? That way we also test the different styles of constructing the issue. Or would that make the tests too verbose?

File: belgif-rest-problem/src/main/java/io/github/belgif/rest/problem/api/InputValidationIssues.java (hunk 8 above)
Comment (jpraet) [reply belgif-rest-problem/src/main/java/io/github/belgif/rest/problem/api/InputValidationIssues.java:116]: It's a standardized issue type at Belgif (https://www.belgif.be/specification/rest/api-guide/#rule-req-valid) which we didn't use at CBSS so far.

I added the corresponding InputValidationIssues.unknownInput() method and a unit test.

File: belgif-rest-problem/src/test/java/io/github/belgif/rest/problem/api/InputValidationIssuesTest.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (jpraet) [reply belgif-rest-problem/src/test/java/io/github/belgif/rest/problem/api/InputValidationIssuesTest.java:14]: I have added assertions to verify the empty properties:

```java
 assertThat(issue).extracting("href", "instance", "status", "inputs", "additionalProperties")
        .allMatch(this::isEmpty);
```

**Review by clone1612 — APPROVED** [review]

**Pull request merged**